dvc/repo/checkout.py

```diff
@@ -59,6 +68,33 @@ def _build_out_changes(index, changes):
     return out_changes
 
 
+def _check_can_delete(
+    entries: List["DataIndexEntry"],
+    index: "BaseDataIndex",
+    path: str,
+    fs: "FileSystem",
+):
+    entry_paths = []
+    for entry in entries:
+        try:
+            cache_fs, cache_path = index.storage_map.get_cache(entry)
+        except ValueError:
+            continue
+
+        if cache_fs.exists(cache_path):
+            continue
+
+        entry_paths.append(fs.path.join(path, *entry.key))
+
+    if not entry_paths:
+        return
+
+    raise DvcException(
+        "Can't remove the following unsaved files without confirmation. "
```

Comment:
Open for suggestions on the phrasing, but this is better than it was and we need this PR to move forward with other stuff, so I'll address it on top if need be.